doc/design/state-machine.md

```diff
@@ -0,0 +1,185 @@
+# HiveD State Machines
+
+This document presents the state machines of HiveD, and explains the life cycles of jobs and resources in HiveD.
+We will describe the state machines of our scheduling unit, Affinity Group, and that of our resource unit, Cell, respectively.
+
+## Affinity Group (AG) State Machine
+
+An affinity group (AG) is a set of gang-scheduled pods. It is the basic scheduling unit in HiveD. The figure below shows the state machine of an affinity group.
+
+Note that the AG state machine has interactions with the cell state machine (elaborated later). In our design, AGs influence each other only via cell: an event of an AG may trigger an event for one of its cells, and the influenced cell may trigger another event for another AG related to the same cell.
+Therefore, the state machines of multiple AGs are effectively bridged by the state machines of their overlapping cells.
+
+![AG](AG-state-machine.png)
+
+### States
+
+__Pending__: the AG is waiting to be scheduled.
+
+__Preempting__: the AG is allocated resources but is waiting for the completion of preemptions of other AGs (only after that can it start to run).
+
+__Allocated__: the AG is allocated resources and is running normally.
```

Comment:
Let's not bring word "running" to the description of "Allocated", too confusing. As Hived does not care if the pod is really running, it only cares resource allocate to the pod, that's end.